### PR TITLE
Assert if we try to use link-time setting before link time. NFC.

### PR DESCRIPTION
**Pull request by sbc100**
We should not be reading or writing to linker-only settings until after
the compile phase.  This change keeps us honest and allows us to be sure
the list of compile time settings that we have is correct (or at least
sufficient, it could contain false positives).

### CODE REVIEW

**Comment by sbc100** [comment]
PTAL.. given that no test fail this should be pretty save to land I think

File: tools/ports/__init__.py [hunk 1]
```diff
@@ -27,6 +27,8 @@ def read_ports():
       assert hasattr(port, a), 'port %s is missing %s' % (port, a)
     if not hasattr(port, 'process_dependencies'):
       port.process_dependencies = lambda x: 0
+    if not hasattr(port, 'linker_setup'):
+      port.linker_setup = lambda x, y: 0
```
Comment (kripken) [inline tools/ports/__init__.py:31]: is this related to the PR?

Comment (sbc100) [reply tools/ports/__init__.py:31]: Yes.  Previously ports were setting link-time flags as part of `process_dependencies`.. which happens during compile time.

This gives them second hook they can use for setting link time flags.   Without this change we would get assertion errors when the ports tries to set link-time flags during the compile phase.

Comment (sbc100) [reply tools/ports/__init__.py:31]: You can see it was just couple of ports that were messing with linker settings.   This change means they won't mess with those settings in `-c` calls..  making `-c` calls simpler and faster.

**Review by kripken — APPROVED** [review]
I see, thanks, lgtm!
